### PR TITLE
#60750 Added Composable  Index Template Builder

### PR DESCRIPTION
**Pull request by antrix190**
**Description**
Added Builder for ComposableIndexTemplate to make it easier to be able to build the template objects. 

**Checks**

- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?  **YES**
- Have you followed the [contributor guidelines](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md)?  **YES**
- If submitting code, have you built your formula locally prior to submission with `gradle check`?  **YES**
- If submitting code, is your pull request against master? Unless there is a good reason otherwise, we prefer pull requests against master and will backport as needed. **YES**
- If submitting code, have you checked that your submission is for an [OS and architecture that we support](https://www.elastic.co/support/matrix#show_os)?  **YES**
- If you are submitting this code for a class then read our [policy] (https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md#contributing-as-part-of-a-class) for that.

### CODE REVIEW

**Comment by antrix190** [comment]
@dakrone Can you check if the changes are as expected. I will complete the TODOs for this PR after you approved the changes.

**Comment by pgomulka** [comment]
@antrix190 please do update the description in this PR

**Comment by antrix190** [comment]
Sure

**Comment by antrix190** [comment]
@pgomulka Done.

**Comment by pgomulka** [comment]
@antrix190 the PR itself should describe the code change and reasoning behind it.

**Comment by antrix190** [comment]
Oh okay my bad. I'll just do that.

**Comment by antrix190** [comment]
@pgomulka I have added an appropriate description.

File: server/src/main/java/org/elasticsearch/cluster/metadata/ComposableIndexTemplate.java [hunk 1]
```diff
@@ -294,4 +299,55 @@ public int hashCode() {
...
+        	return this;
+        }
+        
+        public Builder dataStreamTemplate(DataStreamTemplate dataStreamTemplate) { 
+        	this.dataStreamTemplate = dataStreamTemplate;
+        	return this;
+        }
+        
+        public ComposableIndexTemplate build() { 
+        	return ComposableIndexTemplate(this);
```
Comment (swallez) [inline server/src/main/java/org/elasticsearch/cluster/metadata/ComposableIndexTemplate.java:350]: The code doesn't compile, that should be `return new ComposableIndexTemplate(this)`. Also as mentioned above we should use one of the existing constructors.

File: server/src/main/java/org/elasticsearch/cluster/metadata/ComposableIndexTemplate.java [hunk 2]
```diff
@@ -104,6 +104,11 @@ public ComposableIndexTemplate(List<String> indexPatterns, @Nullable Template te
         this(indexPatterns, template, componentTemplates, priority, version, metadata, null);
     }
 
+    public ComposableIndexTemplate(Builder builder) {
```
Comment (swallez) [inline server/src/main/java/org/elasticsearch/cluster/metadata/ComposableIndexTemplate.java:107]: If you look at other uses of the builder pattern, their `build` method calls one of the regular constructors of the class. This avoids having to add an additional constructor just for the builder.

**Review by swallez — CHANGES_REQUESTED** [review]
The code does not compile (missing `new`) and uses tabs while we use spaces for indentation.

These issues are shown with `./gradlew precommit` as indicated in [the testing doc](https://github.com/elastic/elasticsearch/blob/master/TESTING.asciidoc#running-verification-tasks). Can you fix them and update the PR?

Also left some comments about the new constructor.

File: server/src/main/java/org/elasticsearch/cluster/metadata/ComposableIndexTemplate.java (hunk 1 above)
Comment (antrix190) [reply server/src/main/java/org/elasticsearch/cluster/metadata/ComposableIndexTemplate.java:350]: Done.

File: server/src/main/java/org/elasticsearch/cluster/metadata/ComposableIndexTemplate.java (hunk 2 above)
Comment (antrix190) [reply server/src/main/java/org/elasticsearch/cluster/metadata/ComposableIndexTemplate.java:107]: Agreed. Reused existing constructor.

**Comment by antrix190** [comment]
@swallez requesting your review 🙂

**Comment by swallez** [comment]
test this please

**Comment by antrix190** [comment]
@swallez can you have a look?

**Comment by antrix190** [comment]
hey @swallez can you please have a look?

**Comment by swallez** [comment]
@antrix190 I have re-read our [contribution guidelines](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md#reviewing-and-accepting-your-contribution) which I invite you to read carefully. This PR is mostly a cosmetic refactoring and we tend to only accept contributions that add functionality, fix bugs, or refactor code in service of adding or fixing functionality.

Given the above I am going to close this PR but we would welcome further substantial contributions you would like to propose. One way to find ways to contribute is to look at issues labeled [good first issue](https://github.com/elastic/elasticsearch/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22+sort%3Aupdated-desc) or [help wanted](https://github.com/elastic/elasticsearch/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22+sort%3Aupdated-desc).

**Comment by antrix190** [comment]
@swallez appreciate your opinion and comment. But I worked according to issue #60750 and don't know why it Can't be accepted as a contribution.

**Comment by antrix190** [comment]
@dakrone @swallez its disappointing how after spending time on the issue, I have been told that the issue is only cosmetic change and won't be accepted when the issue is still open.

**Comment by dakrone** [comment]
@swallez can you explain why you feel like this is a cosmetic refactoring? I originally wanted to add a builder but just never had time to get to it, especially as building composable index templates can be pretty verbose. I think we should add a builder (especially since we can make use of it in our rest tests).

**Comment by swallez** [comment]
Hmm... This was following a conversation with @pgomulka who reminded me about our contribution guidelines. I missed the relation to #60750 in the PR title. So let's reopen it and bring it to completion. Sorry for the misunderstanding.

**Comment by antrix190** [comment]
@swallez awesome! can you please have look at the changes.

**Comment by swallez** [comment]
@elasticmachine ok to test

**Comment by swallez** [comment]
@elasticmachine update branch

**Comment by swallez** [comment]
@antrix190 looks like we have quite some flaky tests that prevent the build to happen. We'll get there...

**Comment by antrix190** [comment]
Yeah I see, is there a way to rerun a specific build task, as a hack I had to close and reopen the PR to trigger the build.

**Comment by swallez** [comment]
Let me handle this, I'll check other build results to see if it's an ongoing problem that some other team members are working on.

**Review by swallez — APPROVED** [review]
LGTM

**Comment by swallez** [comment]
@elasticmachine update branch

**Comment by swallez** [comment]
Green build! We got there! Thanks @antrix190!

**Comment by antrix190** [comment]
Thanks for your help @swallez 😁

**Pull request merged**